### PR TITLE
Sans657 findcenterofmassposition2 cleanup fix

### PR DESCRIPTION
**Pull request by walshmm**
**Description of work.**
PR addresses [this defect](https://code.ornl.gov/sns-hfir-scse/sans/sans-backend/-/issues/657)

***Pasted description from the above issue for those who do not have access***:

***==========================================================***

Problem Description:
====================
The current version of [FindCenterOfMassPosition2_wth.cpp](/uploads/18554bbb125f29312fe244db9a4de2e2/FindCenterOfMassPosition2_wth.cpp) contains various

* outdated comments that can are contradictory to the documentation and its actual usage
* dormant bugs resulting in blocks of code [never getting executed](https://github.com/mantidproject/mantid/blob/e821550a4738411be3619f249c1d9fad65262055/Framework/Algorithms/src/FindCenterOfMassPosition2.cpp#L225)
  - Double type vars should not be compared with `==`
  - Also, the convergence criteria here does not make any sense as this is not an iterative method with current implementation.
  - This [block](https://github.com/mantidproject/mantid/blob/e821550a4738411be3619f249c1d9fad65262055/Framework/Algorithms/src/FindCenterOfMassPosition2.cpp#L231) will never get executed
* [satellite block](https://github.com/mantidproject/mantid/blob/e821550a4738411be3619f249c1d9fad65262055/Framework/Algorithms/src/FindCenterOfMassPosition2.cpp#L157) of code the purpose of which is difficult to determine
  - This block should be separated into two parts
    - the first block is locating the physical bounds
    - the second block is to perform the weighted sum (center of mass calculation)
  - Better documentation is needed to correctly reflect how the boundary is adjusted.

The related file on [Github repo](https://github.com/mantidproject/mantid/blob/master/Framework/Algorithms/src/FindCenterOfMassPosition2.cpp)

Steps to Reproduce:
-------------------
Most details are marked as comments in the linked file.

Investigation/Analysis Results:
-------------------------------
* All calculation within the function are carried out in real space units (meters), which need to be reflected in the comments accordingly
* Some `if` block condition cannot be triggered due to incorrect format
* The overall organization of the code can be improved
  * [ ] each block should state their purpose
  * [ ] vars should be declared closed to the start of their scope as a quick summary
  * [ ] move satellite blocks closer to their parents

The changes implemented here can benefits future developers so that they can have a better understanding of the logic behind the code without diving into details.

***============================================***
***Paste End***

- [x] pr against `ornl-next` #32427

Companion to PR ornl-next [GitHub issue/pull request detail](https://github.com/mantidproject/mantid/pull/32427)

**To test:**

Ensure unit tests for `AlgorithmsTest_FindCenterOfMassPosition2Test` still pass

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/inc/MantidAlgorithms/WorkspaceBoundingBox.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+namespace Mantid {
+namespace Algorithms {
+
+/* This is a simple class originally intended for use solely with FindCenterOfMassPosition2.cpp
+ *
+ */
+class WorkspaceBoundingBox {
+public:
+  WorkspaceBoundingBox(API::MatrixWorkspace_sptr workspace, Kernel::Logger &g_log);
+  WorkspaceBoundingBox(Kernel::Logger &g_log);
```
Comment (KedoKudo) [inline Framework/Algorithms/inc/MantidAlgorithms/WorkspaceBoundingBox.h:21]: Is it necessary to passing the logger this way?  You should be able to define a logger in the namespace 

```c++
namespace Mantid {
namespace Algorithms {
    namespace {
    // static logger
    Logger g_log("WorkspaceBoundingBox");
    } // namespace
...
}
}
```

File: Framework/Algorithms/inc/MantidAlgorithms/WorkspaceBoundingBox.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  double centerX;
+  double centerY;
+  double xMin = 0;
+  double xMax = 0;
+  double yMin = 0;
+  double yMax = 0;
+  const int m_specID = 0;
+
+  /// Logger for this helper
+  Kernel::Logger &g_log;
```
Comment (KedoKudo) [inline Framework/Algorithms/inc/MantidAlgorithms/WorkspaceBoundingBox.h:62]: Please consider using the previous suggested method to give this logger a name (easier for future trouble shooting).

**Review by KedoKudo — CHANGES_REQUESTED** [review]
Please add documentation to methods with arguments. Here is an example below:
```c++
/**
 *  Get the weight factor that would be used for an event occuring
 *  at the specified wavelength, with the specified two_theta value.
 *
 *  @param  lamda      The wavelength of an event.
 *  @param  two_theta  The scattering angle of the event.
 *
 *  @return The weight factor for the specified position and wavelength.
 */
double AnvredCorrection::getEventWeight(double lamda, double two_theta) {
```
This will help the doc generator to auto generate useful doc pages as well as the IntelliSense from VSCode (last one is more of a personal request).

File: Framework/Algorithms/inc/MantidAlgorithms/WorkspaceBoundingBox.h (hunk 1 above)
Comment (walshmm) [reply Framework/Algorithms/inc/MantidAlgorithms/WorkspaceBoundingBox.h:21]: This helper class is tied pretty strongly to FindCenterOfMass2, as such I thought it should just continue to use the FindCenterOfMass2 logger as not to confuse the log readers.  I can change it if you still think it is necessary.

**Comment by walshmm** [comment]
rebased off of release-next. hopefully didnt break anything

File: Framework/Algorithms/inc/MantidAlgorithms/WorkspaceBoundingBox.h (hunk 1 above)
Comment (KedoKudo) [reply Framework/Algorithms/inc/MantidAlgorithms/WorkspaceBoundingBox.h:21]: I think it's better to let users know where the messages actually coming from.
Please make `WorkspaceBoundingBox` using its own logger :pray:

Comment (walshmm) [reply Framework/Algorithms/inc/MantidAlgorithms/WorkspaceBoundingBox.h]: alrighty, added

**Review by KedoKudo — DISMISSED** [review]
looks good :+1:

**Review by gvardany — CHANGES_REQUESTED** [review]
I am not able to open the link describing the defect, as I guess it's ornl only. Please copy the description to the body of the PR.

**Comment by walshmm** [comment]
> I am not able to open the link describing the defect, as I guess it's ornl only. Please copy the description to the body of the PR.

@gvardany I have pasted the description as requested.

File: Framework/Algorithms/inc/MantidAlgorithms/WorkspaceBoundingBox.h [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  double centerY;
+  double xMin = 0;
+  double xMax = 0;
+  double yMin = 0;
+  double yMax = 0;
+  const int m_specID = 0;
+};
+
+} // namespace Algorithms
+} // namespace Mantid
```
Comment (gvardany) [inline Framework/Algorithms/inc/MantidAlgorithms/WorkspaceBoundingBox.h:107]: There is a missing new line at the end.

File: Framework/Algorithms/inc/MantidAlgorithms/FindCenterOfMassPosition2.h [hunk 4]
```diff
@@ -63,6 +65,36 @@ class MANTID_ALGORITHMS_DLL FindCenterOfMassPosition2 : public API::Algorithm {
...
+  /// Helper functions
+  /** Iterates through spectrum in the input workspace finding the center of mass until
+   *  we converge to within the tolerance specified in meters
+   *
+   *  @param inputWS  :: workspace to find the center of mass of
+   *  @param centerX  :: save the real center of mass x coord here
+   *  @param centerY  :: save the real center of mass y coord here
+   *  @param numSpec  :: number of spectrum in the workspace to iterate through
+   *  @param progress :: object for reporting progress of the operation
+   */
```
Comment (gvardany) [inline Framework/Algorithms/inc/MantidAlgorithms/FindCenterOfMassPosition2.h:77]: I think the doxygen comments should go to cpp and not the header.

**Pull request merged**